hnn_core/parallel_backends.py

```diff
@@ -282,67 +304,51 @@ def simulate(self, net):
+                          self._read_stdout)
+        self.sel.register(pipe_stderr_r, selectors.EVENT_READ,
+                          self._read_stdout)
+
+        # loop while the process is running
         while proc.poll() is None:
-            ready = select.select([pipe_stdout_r, pipe_stderr_r], [], [], 0)[0]
-            if not ready:
-                continue
-            for stream in ready:
-                if stream == pipe_stdout_r:
-                    buf = os.read(pipe_stdout_r, 1024)
-                    # write processes stdout to our stdout (fd 0)
-                    try:
-                        stdout_fd = sys.stdout.fileno()
-                    except AttributeError:
-                        # for sphinx-gallery that replaces sys.stdout
-                        stdout_fd = 0
-                    os.write(stdout_fd, buf)
-                else:
-                    proc_stderr_bytes += os.read(pipe_stderr_r, 1024)
+            # wait for an event on the selector, timeout after 1s and check
+            # that process is still running
+            events = self.sel.select(timeout=1)
+            for key, mask in events:
+                callback = key.data
+                finished = callback(key.fileobj, mask)
+                if finished:
+                    # finishied receiving printable output
+                    self.sel.unregister(pipe_stdout_r)
+                    self.sel.unregister(pipe_stderr_r)
+                    # everything received should now be handled by _read_data
+                    self.sel.register(pipe_stderr_r, selectors.EVENT_READ,
+                                      self._read_data)
+
+        self.sel.unregister(pipe_stderr_r)
+        self.sel.close()
 
         # done with stdout and stderr
         os.close(pipe_stdout_r)
         os.close(pipe_stdout_w)
         os.close(pipe_stderr_r)
         os.close(pipe_stderr_w)
 
-        # covert to string to be able to look for padding
-        proc_stderr = proc_stderr_bytes.decode()
-
-        sim_failed = False
-        mpi_error_msg = ''
-
-        # data will always be padded with '===' at the end.
-        if "=" in proc_stderr:
-            # see if there are any error messages after data (padded with "==")
-            stderr_list = proc_stderr.split("=")
-            if len(stderr_list[-1]) > 0:
-                mpi_error_msg = stderr_list[-1]
-                data_str = stderr_list[0]  # padding removed
-                data_str_bytes = data_str.encode() + b"==="
-            else:
-                data_str = proc_stderr
-                data_str_bytes = data_str.encode()
-        else:
-            sim_failed = True
-            if len(proc_stderr) > 0:
-                mpi_error_msg = proc_stderr
-
-        if sim_failed or len(mpi_error_msg) > 0:
-            sys.stderr.write("MPI simulation returned extra messages:\n")
-            sys.stderr.write(mpi_error_msg)
-            sys.stderr.write("\n*** End MPI messages ***\n\n")
-
```

Comment:
This was the code to search for padding of the data and separate out error messages/exceptions. Yuck.